### PR TITLE
resolves RPi-Distro/raspberrypi-sys-mods#42: prompts user for third p…

### PR DESCRIPTION
**Pull request by BitBistro-code**
resolves RPi-Distro/raspberrypi-sys-mods#42

Please consider this patch that would prompt users if they would like to install third party repositories. If the vscode repo already exists no attempt will be made to change it regardless to the answer of the question. User should be prompted once, or if they rerun dpkg-reconfigure.

The hooks added will allow future debconf questions to be added relatively easily. It should also be setup for localization, but I can only confidently add English text.

Also, slightly modified the way we write out the key. It is first written to a temp file instead of what ever the current directory of the process is. It is then moved as normal and chmod'd.

### CODE REVIEW

**Comment by mikewwwperry** [comment]
I should note, the default is true: add the repo. The user can easily preload debconf with the response. There are tutorials online on how to do this, but I could write something up if needed.

**Review by RubenNL — APPROVED** [review]

File: debian/templates
Comment (c33s) [inline debian/templates:4]: as you already stated in your comment https://github.com/RPi-Distro/raspberrypi-sys-mods/pull/51#issuecomment-774732103 the default is a `yes`. personally i think this should be a `no`.

i am not sure but i assume that running `apt-get --yes install ...` the `--yes` will also answer this question with `yes` will it? (haven't tested). if that is true the default should be `no`. 

running apt with `--yes` is often done to automated bootstrap systems, like for docker images, for use with puppet and so on.
from GDPR view the default also should not (must not?) be `yes` as the ip of the users computer is sent to microsoft without knowing. an ip address counts as personal information. tracking must be opt in and not opt out. the yes from install the package `raspberrypi-sys-mods` is not the yes for install a 3rd party repo. compareable with java license agreement or lets encrypt agreement. if you want to install them automated you have to set an environment variable or provide a specific parameter.

Comment (mikewwwperry) [reply debian/templates:4]: As I am not a normal maintainer for this repository, I did not feel that I should change the default behavior. I will leave it up to the true maintainers to modify the default if they accept my PR.

The mechanism I am using here to prompt the user (or not for that matter) is debconf. Debconf is well integrated into the Debian ecosystem and you have probably already used it without knowing. It is the same thing that is used to select your language, keyboard, and timezone among other things. These choices can be preloaded, and persisted across updates/installs (unless the package is purged). 

Under normal circumstances, Debconf will prompt the user if:

- They haven't seen the question before / they haven't already added in the value ahead of time
- The priority of the question is at or above the priority they wish to see (low, medium, high, critical)
- Debconf can interact with their specified front-end

If cases  the user is not prompted, then the default is used.

You can see the choices of frontends you have available by running:
`sudo dpkg-reconfigure -plow debconf`
Your current selections for debconf should already be set at their current value.

So to specify your own default, you would just need to preload the debconf database. For this change you could use a command something like this:
`echo "raspberrypi-sys-mods raspberrypi-sys-mods/allow_3p_repos select no" | sudo debconf-set-selections`

The trick here is that you would have to populate this information prior to installing the packing. However, if you are running interactively, allow prompting of medium priority questions, and do not preconfigure the value you will still be prompted anyway.

Software designed to allow building of custom Raspberry Pi images should already have hooks for debconf. In fact, the offical builder, [pi-gen](https://github.com/RPi-Distro/pi-gen), documents that hook right on the readme (Search for 00-debconf).

I am not aware of any GDPR requirement that this package would be in breach of (although I am not a lawyer). There are certain compliance agencies, such as NERC, that are very strict. In such a case, they would not be using stock images unless they were heavily vetted. All changes would be going through strict change control. 

For general usage, I feel that debconf is a excellent tool to give end users and administrators the ability to modify package behavior. As I said, I believe the upstream software maintainer(s) should choose the default value. This code just makes it easier to provide that choice.

Comment (c33s) [reply debian/templates:4]: @mikewwwperry it is all about changing the default behavior not keeping it in the trojan state :)

as i am from europe i am quite often confronted with GDPR. i am also no lawyer but what i can say (which a lawyer can hopefully confirm) that tracking have to be opt-in and not opt-out (this is where all this terrible cookie banners come from).

some somple GDPR facts for developers:
- having the M$ repo installed means the user pings M$, they get the users ip -> tracking
- tracking must be opt-in
- default has to be `no`

it is not about allowing the user/developer/builder to disable the tracking by opt-out via a env var or configure by preconfiguring debconf. it is a privacy first requirement. so if someone wants to build an image with the ms repo, then they have to actvily enable it with the debconf tool

i fully understand that you are no core developer and i thank you for your time and effort providing this PR without knowing if it even will be merged.
with my comment i just want to make certain that the maintainers may have a real problem with GDPR in the current repos state and maybe also after they merge this if they don't change the default setting.

Comment (FredericGuilbault) [reply debian/templates:4]: At this point I think that, yes it should use debconf to ask user contentment. But also in ti's own package as t'its quite an optional `raspberrypi-sys[tem]-modI[ification]s`. 

A nice suggested package called something like vendors_repo.deb or non-free_corporate_repos.deb. 

Anyway, thx for submitting something without knowing if they even plan to fix it.

**Review by FredericGuilbault — APPROVED** [review]

File: debian/po/templates.pot
Comment (MichaIng) [inline debian/po/templates.pot:30]: Does it make sense to allow this generally? I think a 3rd party repository should not added at all by an APT package, unless you explicitly choose to install the software that is provided by this 3rd party repository.

Since the decision about the VS Code topic has been done already, at least a debconf choice should not grant now to add unlimited further repos. Instead this should be a very rare (or the only) exception and if for some reason the question is to add another 3rd party repo, this should require an own debconf question for this specific new repository.

Comment (mikewwwperry) [reply debian/po/templates.pot:30]: This is just a merge of the changes that @XECDesign has already pull into the HEAD to prevent conflicts. This PR, 51, does provide user with choice as explained above:
https://github.com/RPi-Distro/raspberrypi-sys-mods/pull/51#discussion_r571704423

**Comment by MichaIng** [comment]
Sorry for doing a code comment, this was actually not the right place since it is a general concern about the aim of this PR:
- This PR implements a debconf authorisation for having third party repositories added. Do not understand me wrong, debconf is probably the best way to deal with such things. 
- But I suggest to implement a debconf authorisation for having the Microsoft VS Code repository added only.
- As this aims to handle future 3rd party repository implementations (which at best should never be required), I think those should get a very own debconf from for every single repo, stating full details about the repo that is to be added and why maintainers think it is helpful/important enough to even bother with it.
- So an authorisation for adding all future 3rd party repos silently is not a choice that should exist, but only for adding a specific individual repo one can review first, and decide whether useful (or harmful) or not. 

Since the VS Code repo has been added now for days already, and the postinst code won't add it a second time, it is probably not effective for that matter anymore and will only force an unnecessary debconf for a decision that may have been made already (I removed the repo and key, decided to do so, with this PR I'm asked again about this matter), but it serves as a template and reminder about how such may be done for 3rd party repositories in the future in the first place.

File: debian/templates
Comment (jpmcneely) [reply debian/templates:4]: This is the right way - a new package.

**Comment by mikewwwperry** [comment]
Thank you everyone for the comments. I moved the default to no, and set the priority to high. Regardless of what I set these to the maintainers can change as they see fit; so I suppose I have defeated my own argument. From my standpoint it seems like this is what the community would prefer.

@MichaIng - No problem about the comments. I guess you could comment directly at my fork or just email me. I too feel like this type of functionality might be better for some other package. Personally, I was thinking raspi-config or integrated into the recommend software desktop application. That said, my goal was just provide a better method given the current implementation. I'm a true believer in iterative development. It was put in this package for some reason. It could still be moved.

Also, @MichaIng, I don't think it is too late. Sure a lot of people have gone through and already added this source. A lot have not. Hopefully, there are also many more future installs of Raspberry PI. So even though this isn't a retroactive fix, it is in place for posterity.

I guess I never stated why I did this. It actually isn't about Microsoft, but about user choice. I don't think a package should pull in **any** third party repository without user consent. 

Why isn't it retroactive or bigger? This is just a hobby for me. I have a day job and a family. As it is, I had to learn how to use debconf and build this package. It was done with love for the community and the Pi. 

@XECDesign - Can you please comment on this PR. I am interested in your thoughts as well, if you would like me to invest more time in it some way. Or if you feel I should hang up my hat on this and call it day. If you want to keep it off thread, please email me; I'll keep it confidence unless you say otherwise.

Mike Perry <mike@bitbistro.org>

**Comment by MichaIng** [comment]
> Hopefully, there are also many more future installs of Raspberry PI. So even though this isn't a retroactive fix, it is in place for posterity.

Just to clarify that point. The current upstream code does not install the repository on a fresh package install. Only package upgrades from existing pre-20210125 installs will have the repo installed. This even means that with current upstream pi-gen code, creating a fresh Raspberry Pi OS image would result in this repo never being added. So this PR makes pi-gen users and such, who went through the upgrade already, being faced with the repo again, while currently they wouldn't . This is no killer argument, since a transparent way of adding such a repo, including a way to chose against it once and forever, interactively and non-interactively (path-exclude, debconf-set-selection, ...) is much less of an issue. But it is a question whether one wants to (re)awaken the topic for users where it is no topic (anymore), especially after the massive negative and heated feedback this caused, where probably even a good way of dealing with it might wake negative emotions 😅.

At the day the upgrade arrived I was about to write a commit to make it a regular config file, which is added transparently and can easily be blocked and custom changes/removal are persistent unless one uses DPkg force-confmiss/new options. After recognising the above fact, I decided against finishing the commit, as long as the postinst code keeps behaving like it does now.

However, the more I think about it, the more I agree that debconf indeed would have been the best way, so even if it makes e.g. my systems facing another prompt, unless I pro-actively apply `debconf-set-selection`, it ads a great template that can be used for future sensitive changes as well. In theory it should be even possible to automatically answer the question with "no", if the package version is >=20210125 and the repo does not exist?

Btw:
> If the vscode repo already exists no attempt will be made to change it regardless to the answer of the question.

Actually, given that there are likely users who didn't even recognise it, IMO it would make sense to ask, (especially) if the repo exists already and no answer has been given yet, and to remove it if the answer is "no". If the repo does not exist, but the package version is >=20210125 already, it must have been removed manually, hence the question is obsolete and "no" can be assumed. That way, all my above concerns would be dealt 🙂.

However, I then still think instead of:
```
yes: new third party repositories may be added
```
it should be :
```
yes: the Microsoft VS Code third party repository may be added
```

But let's see what maintainer(s) thing of the general idea to use debconf.

**Comment by BitBistro-code** [comment]
@MichaIng - I didn't have an opportunity to do a fresh install and see the behavior. If the fresh install starts with a package prior to the 20210125 version, then I believe the repo would be added. I'd have to run some tests to see exactly how `dpkg --compare-versions "${2}" lt-nl "20210125"` would evaluate. I think this will evaluate true for older versions or empty version.

Have you run a fresh install to see the behavior w/out my patch? I have a new Pi I'm building so I might be able to do some testing soon. Time is a resource. If you can say for certain that a fresh install after an update does not have the repo then I can change my branch to include the dpkg compare check. Then behaviors should match. Although I did attempt to change my default here, and there is of course the preload option. But yes, I agree making not a topic to argue about would be good.

I tried to make it not specific to the Microsoft Repo and call it third party. If I am going to call out the specific repo, then it should be a multi-select like ca-certs uses. I was trying to keep it simple. I can improve up on this if we decide to move forward. I'm relucant to put any more effort into it until I hear further. If that is the case too, lets make sure that this package is the right spot.

**Comment by MichaIng** [comment]
`"${2}" lt "20210125"` would evaluate true for empty versions.
`"${2}" lt-nl "20210125"` evaluates false for empty versions.

Check out the man page (search for "--compare"): https://manpages.debian.org/dpkg
And you don't require a new system to test it, purge the package and install it again 😉.

> If the fresh install starts with a package prior to the 20210125 version, then I believe the repo would be added.

Ah that is true, so it's installed with the current Raspberry Pi OS images from January 11. I'm still waiting to the `pi-gen` patch to have it added for new RPi OS builds as well.

> I tried to make it not specific to the Microsoft Repo and call it third party. If I am going to call out the specific repo, then it should be a multi-select like ca-certs uses. I was trying to keep it simple.

If further repos are about to be added, I agree a multi-select would be better than multiple yes/no prompts, as long as there is enough space for some info about each repo.
It's just that I would never hit "yes" on a question whether generally the package is allowed to add 3rd party repos, unattended from now on, but I want to be able to decide it for each repo individually. If I like to use VS Code, okay, but that doesn't mean that I need or want the Docker repository added automatically in the future, let's say. Currently the questions sounds like it would permit exactly that.

**Comment by BitBistro-code** [comment]
@MichaIng Well that was an interesting endevor. I forgot that updating my branch will automatically update the PR. Good lesson., but I do love to make commits. Anyway, I tested the behavior and with or without the dpkg compare check the user is prompted for new installs without the package existing. However, I believe most pi distros will start with some base version, which will either add or not add the repos depending on their version/date.

However I ran into some poor behavior where dpkg-reconfigure will not add the repo if the user the user changes their answer from no to yes. Then I remembered that was why I removed the compare in the first place.

So in order to cast the widest net, and I must say it is impossible to make everyone happy, I added a purge option. So the options are now as follows (with default no):

  - yes: new third party repositories may be added;
  - no: new third party repositories will not be added;
  - purge: new third party repositories will not be added, and previously add third party repositories will be purged;

When purging I do a very rudamentary attempt at seeing if they changed the files. If they did, I do not purge. I use cksum, which is horrible for security implications, but good enough to do this job. It should be installed even on very barebones systems as it is in coreutils which has a required priority. It also should not tax the cpu much. These files are small anyway.

So I said that was it unless I hear more from the maintainers and this is probably it.

**Comment by XECDesign** [comment]
Since our aim is to remove hurdles and make things simple for beginners who may not know what a repo is, requiring input like that is just confusing.

I can see that you've put time in this commit and it's a great example of how to use debconf, but it doesn't fit with what we're going for, I'm sorry.

**Comment by robertoschwald** [comment]
> I can see that you've put time in this commit and it's a great example of how to use debconf, but it doesn't fit with what we're going for, I'm sorry.

Interesting. What _are_ you going for? Own m$ package, leave _as_is_, ... ?

**Comment by XECDesign** [comment]
We'll see.

**Comment by MichaIng** [comment]
> make things simple for beginners who may not know what a repo is

The question can be rephrased as it is simply about whether they what to be able to more easily install Microsoft VS Code or not. If someone does not know what this is, or simply does not want to use it, one does not need to understand what a repo is, but can fully qualified answer "Thanks, not needed", that's all.

**Comment by XECDesign** [comment]
Sure and maybe that will be something we might do, but there are too many questions up in the air right now.

**Comment by BitBistro-code** [comment]
@XECDesign I know your didn't ask for this so I wasn't sure what to expect. Coding is in my blood and I've been cooped up at home because of covid I've needed outlets.

After going down this path my recommendation is to at least add a way to raspi-config to remove the repo. That may be a better place for the add to live to.

There is also the gui package that does the recommended software packages. That depends on things like package-kit that should be able to help handle adding and removing the repo. That could be even better because I don't think lite versions of the Pi even have that package.

Finally, please reach out to me if you want some help with something @XECDesign. I'm not sure what the protocol is for nmu pull requests. I ran into a few other oddities though in other packages and can probably put in some time for other fixes.

**Comment by XECDesign** [comment]
Thank you very much for your kind approach.

When it comes to this particular issue, at this point, all options are on the table, including removing the repo entirely. That's the only reason I can't give any helpful comments or accept any pull requests right now - it may all become moot.

Contributions are certainly welcome, but they require review so I find that sometimes issues and PRs hang around much longer than they would in the ideal world. I wouldn't want somebody to send a pull yet another pull request and then have it hang for years because of higher priority tasks.

Maybe open issues for the oddities you've spotted over at https://github.com/RPi-Distro/repo/ and then we can go from there?

In general if it's something like a lintian warning or something that has no practical downsides, it's near the bottom of the list. If it's something that affects niche use cases or the PR is particularly complicated and has potential to mess things up elsewhere, it may take a while to get to as well.

**Pull request closed without merging**
